src/types.py

```diff
@@ -930,6 +943,112 @@ def ctype(ctype: CType, typemap: TypeMap, typepool: TypePool) -> "Type":
             return Type(TypeData(kind=TypeData.K_INT, size_bits=size_bits, sign=sign))
         static_assert_unreachable(real_ctype)
 
+    @staticmethod
+    def demangled_symbol(sym: CxxSymbol) -> "Type":
+        return Type.demangled_type(sym_type=sym.type, sym_name=sym.name)
+
+    @staticmethod
+    def demangled_type(sym_type: CxxType, sym_name: Optional[CxxTerm] = None) -> "Type":
+        # TODO: This function may need to depend on the ABI, so it may need to be moved out of types.py?
+        final_name = None
+        if sym_name is not None:
+            assert sym_name.kind == CxxTerm.Kind.QUALIFIED
+            assert sym_name.qualified_name is not None
+            assert len(sym_name.qualified_name) >= 1
+            final_name = str(sym_name.qualified_name[-1]).rpartition("@")[-1]
+
+        type = None
+        for term in sym_type.terms[::-1]:
+            if term.kind == CxxTerm.Kind.CONST:
+                pass
+            elif term.kind == CxxTerm.Kind.UNSIGNED:
+                unsigned = True
+            elif term.kind in (CxxTerm.Kind.POINTER, CxxTerm.Kind.REFERENCE):
+                type = Type.ptr(type)
+            elif term.kind == CxxTerm.Kind.BOOL:
+                type = Type.bool()
+            elif term.kind == CxxTerm.Kind.CHAR:
+                type = Type.int_of_size(8)
+            elif term.kind in (CxxTerm.Kind.SHORT, CxxTerm.Kind.WIDE_CHAR):
+                type = Type.int_of_size(16)
+            elif term.kind in (CxxTerm.Kind.INT, CxxTerm.Kind.LONG):
+                type = Type.int_of_size(32)
+            elif term.kind == CxxTerm.Kind.LONG_LONG:
+                type = Type.int_of_size(64)
+            elif term.kind == CxxTerm.Kind.FLOAT:
+                type = Type.f32()
+            elif term.kind == CxxTerm.Kind.DOUBLE:
+                type = Type.f64()
+            elif term.kind == CxxTerm.Kind.LONG_DOUBLE:
+                type = Type.f128()
+            elif term.kind == CxxTerm.Kind.SIGNED:
+                assert type is not None
```

Comment:
I had thought these asserts should be guaranteed by `demangle_codewarrior.py` (`SIGNED`, `UNSIGNED`, `ARRAY`, etc. are all "`NONTERMINATING_KINDS`") -- but it's not actually required.

(The `array_dim` for `ARRAY`s *is* guaranteed though)

Reply:
I'm refactoring this so that `type` is never `None`: it can start as `Type.any()`, and if the last item in `sym_type.terms` is not a `NONTERMINATING_KIND` (i.e. it *is* a terminating kind, like a `CHAR`, `SHORT`, etc.), it'll be replaced.